enterprise_access/apps/provisioning/api.py

```diff
@@ -108,3 +109,59 @@ def get_or_create_enterprise_catalog(enterprise_customer_uuid, catalog_title, ca
     )
     logger.info('Provisioning: created enterprise catalog with uuid %s', created_catalog.get('uuid'))
     return created_catalog
+
+
+def get_or_create_customer_agreement(enterprise_customer_uuid, customer_slug, default_catalog_uuid=None, **kwargs):
+    """
+    Get or create a customer agreement record from the license-manager service.
+    """
+    client = LicenseManagerApiClient()
+    existing_agreement = client.get_customer_agreement(enterprise_customer_uuid)
+    if existing_agreement:
+        logger.info('Provisioning: customer agreement with uuid %s already_exists', existing_agreement.get('uuid'))
+        return existing_agreement
+
+    created_agreement = client.create_customer_agreement(
+        enterprise_customer_uuid,
+        customer_slug,
+        default_catalog_uuid=default_catalog_uuid,
+        **kwargs,
+    )
+    logger.info('Provisioning: created customer agreement with uuid %s', created_agreement.get('uuid'))
+    return created_agreement
+
+
+def get_or_create_subscription_plan(
+    customer_agreement_dict, plan_title, catalog_uuid, opp_line_item,
+    start_date, expiration_date, desired_num_licenses, **kwargs
+):
+    """
+    Get or create a new subscription plan, provided an existing customer agreement dictionary.
+    """
+    existing_subscriptions = customer_agreement_dict.get('subscriptions', [])
+    matching_subscription = next((
+        _sub for _sub in existing_subscriptions if _sub.get('title') == plan_title
+    ), None)
```

Comment:
I think the `opp_line_item` might be a better candidate for de-duplication than the `title` because that's closer to representing the source record of truth which beget this plan.  If we create a new opp_line_item as a result of new spend, but randomly give it a duplicate title, then that either feels like a warning or error situation, not a success.

Reply:
Ah, great point, thank you